### PR TITLE
Initialization of the database only once in Docker.

### PR DESCRIPTION
**Pull request by AlexNik**
### Changelog category:
- Build/Testing/Packaging Improvement

### Changelog entry (a user-readable short description of the changes that goes to CHANGELOG.md):
If the database is already initialized, it doesn't need to be initialized again upon subsequent launches. This can potentially fix the issue of infinite container restarts when the database fails to load within 1000 attempts (relevant for very large databases and multi-node setups).

### CODE REVIEW

**Comment by den-crane** [comment]
wdyt @Felixoid

**Comment by robot-ch-test-poll** [comment]
This is an automated comment for commit 4766f22decb028e9858707af36480971808f95a6 with description of existing statuses. It's updated for the latest CI running
The full report is available [here](https://s3.amazonaws.com/clickhouse-test-reports/50724/4766f22decb028e9858707af36480971808f95a6/ci_running.html)
The overall status of the commit is 🟡 pending

<table><thead><tr><th>Check name</th><th>Description</th><th>Status</th></tr></thead>
<tbody><tr><td>CI running</td><td>A meta-check that indicates the running CI. Normally, it's in <b>success</b> or <b>pending</b> state. The failed status indicates some problems with the PR</td><td>🟡 pending</td></tr>
<tr><td>Mergeable Check</td><td>Checks if all other necessary checks are successful</td><td>🟢 success</td></tr>
</table>

**Comment by Felixoid** [comment]
It makes sense to implement it similarly to [mysql](https://github.com/docker-library/mysql/blob/master/8.0/docker-entrypoint.sh) and [postgresql](https://github.com/docker-library/postgres/blob/master/16/bullseye/docker-entrypoint.sh#L312) images, so it requires some refactoring to make it nices

**Comment by AlexNik** [comment]
@Felixoid is it OK now?
I'va added `DATABASE_ALREADY_EXISTS` variable and shift for if body.

**Review by Felixoid — APPROVED** [review]
looks OK to me, asking for another pair of eyes though

**Comment by qoega** [comment]
BTW do we have at least some trivial tests that use this docker images? Maybe it is time to add some...

**Comment by AlexNik** [comment]
@Felixoid What is next?

File: docker/server/entrypoint.sh [hunk 1]
```diff
@@ -104,66 +104,76 @@ if [ -n "$CLICKHOUSE_USER" ] && [ "$CLICKHOUSE_USER" != "default" ] || [ -n "$CL
...
-        URL="http://127.0.0.1:$HTTP_PORT/ping"
-    else
-        URL="https://127.0.0.1:$HTTPS_PORT/ping"
-    fi
-
-    # Listen only on localhost until the initialization is done
-    /usr/bin/clickhouse su "${USER}:${GROUP}" /usr/bin/clickhouse-server --config-file="$CLICKHOUSE_CONFIG" -- --listen_host=127.0.0.1 &
-    pid="$!"
+# checking $DATA_DIR for initialization
+if [ -d "${DATA_DIR}data" ]; then
```
Comment (Felixoid) [inline docker/server/entrypoint.sh:108]: what if `${DATA_DIR}` doesn't have a trailing slash? I'd rather like to have an odd one, than risk to have `/var/lib/clickhousedata` instead

**Review by Felixoid — COMMENTED** [review]
Hi. I agree with Ilya, we need to add a test for the public image(s). At least for clickhouse-server.

And then it should cover different use cases: start w/o anything; start with a different user; start with applying initdb.

This is quite a significant change, and I struggle to merge it blindly.

There are once again and again some urgent tasks, that distracts me from https://github.com/ClickHouse/ClickHouse/issues/31473, but I think, I'll implement the mentioned tests in the scope of the publishing clickhouse as an official image

File: docker/server/entrypoint.sh (hunk 1 above)
Comment (AlexNik) [reply docker/server/entrypoint.sh]: rewrote to
${DATA_DIR%/}/data"

**Comment by alexey-milovidov** [comment]
@Felixoid last activity in https://github.com/ClickHouse/ClickHouse/issues/31473 was half a year ago.

**Comment by Felixoid** [comment]
1 time-out test for the debug build looks ok, it succeeded everywhere else

**Comment by Alesh** [comment]
Couldn't you have left the previous option for developer mode? It was convenient to immediately add the creation of new entities to the initialization script. Now we have to come up with workarounds.

**Comment by vincentbernat** [comment]
This change is inconvenient if we still need to add some init steps even if the database exists. I am using it to:
- update ClickHouse configuration
- install Protobuf schemas

**Comment by alexey-milovidov** [comment]
Ok, looking for another PR with improvement.

**Pull request merged**